rust/src/smb/smb.rs

```diff
@@ -0,0 +1,1862 @@
+
+    pub ssn2tree_map: HashMap<SMBCommonHdr, SMBTree>,
+
+    // track the max size we expect for TRANS responses
+    pub ssn2maxsize_map: HashMap<SMBCommonHdr, u16>,
+    pub ssnguid2vec_map: HashMap<SMBHashKeyHdrGuid, Vec<u8>>,
+
+    /// TCP segments defragmentation buffer
+    pub tcp_buffer_ts: Vec<u8>,
+    pub tcp_buffer_tc: Vec<u8>,
+
+    pub files: SMBFiles,
+
+    pub skip_ts: u32,
+    pub skip_tc: u32,
+
+    pub file_ts_left : u32,
+    pub file_tc_left : u32,
+    pub file_ts_guid : Vec<u8>,
+    pub file_tc_guid : Vec<u8>,
+
+    pub ts_ssn_gap: bool,
+    pub tc_ssn_gap: bool,
+
+    pub ts_gap: bool, // last TS update was gap
+    pub tc_gap: bool, // last TC update was gap
+
+    pub ts_trunc: bool, // no more data for TOSERVER
+    pub tc_trunc: bool, // no more data for TOCLIENT
+
+    /// transactions list
+    pub transactions: Vec<SMBTransaction>,
+
+    /// tx counter for assigning incrementing id's to tx's
+    tx_id: u64,
+
+    /// SMB2 dialect or 0 if not set or SMB1
+    pub dialect: u16,
+    /// contains name of SMB1 dialect
+    pub dialect_vec: Option<Vec<u8>>, // used if dialect == 0
+
+    /// dcerpc interfaces, stored here to be able to match
+    /// them while inspecting DCERPC REQUEST txs
+    pub dcerpc_ifaces: Option<Vec<DCERPCIface>>,
+}
+
+impl SMBState {
+    /// Allocation function for a new TLS parser instance
+    pub fn new() -> SMBState {
+        SMBState {
+            ssn2vec_map:HashMap::new(),
+            guid2name_map:HashMap::new(),
+            ssn2vecoffset_map:HashMap::new(),
+            ssn2tree_map:HashMap::new(),
+            ssn2maxsize_map:HashMap::new(),
+            ssnguid2vec_map:HashMap::new(),
+            tcp_buffer_ts:Vec::new(),
+            tcp_buffer_tc:Vec::new(),
+            files: SMBFiles::new(),
+            skip_ts:0,
+            skip_tc:0,
+            file_ts_left:0,
+            file_tc_left:0,
+            file_ts_guid:Vec::new(),
+            file_tc_guid:Vec::new(),
+            ts_ssn_gap: false,
+            tc_ssn_gap: false,
+            ts_gap: false,
+            tc_gap: false,
+            ts_trunc: false,
+            tc_trunc: false,
+            transactions: Vec::new(),
+            tx_id:0,
+            dialect:0,
+            dialect_vec: None,
+            dcerpc_ifaces: None,
+        }
+    }
+
+    pub fn free(&mut self) {
+        //self._debug_state_stats();
+        self._debug_tx_stats();
+        self.files.free();
+    }
+
+    pub fn new_tx(&mut self) -> SMBTransaction {
+        let mut tx = SMBTransaction::new();
+        self.tx_id += 1;
+        tx.id = self.tx_id;
+        SCLogDebug!("TX {} created", tx.id);
+        return tx;
+    }
+
+    pub fn free_tx(&mut self, tx_id: u64) {
+        SCLogDebug!("Freeing TX with ID {} TX.ID {}", tx_id, tx_id+1);
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
```

Comment:
You can use an iterator here. For ex, here is an idiomatic form:
``
// and_then will call the function only if a transaction was found
self.transactions.iter()
    .position(|ref tx| tx.id == tx_id + 1)
    and_then(self.transactions.remove);
``
(sorry for the indentation, I lost the fight with github's markdown)

Reply:
Sorry, correct syntax would be:
```
    self.transactions.iter()
            .position(|ref tx| tx.id == tx_id + 1)
            .map(|idx| self.transactions.remove(idx));
```

Reply:
how would you keep the debug statement, logging the tx to be removed, in this?